### PR TITLE
MAINT Use class_of_interest in DecisionBoundaryDisplay

### PR DESCRIPTION
**Pull request by ArturoAmorQ**
Since scikit-learn v1.4 `DecisionBoundaryDisplay` accepts `class_of_interest` for multiclass visualization. This feature is promised in the current version of the MOOC.

Notice that, as it requires updating the minimal version, it may change the experience of current enrolled participants.

### CODE REVIEW

File: python_scripts/trees_sol_01.py [hunk 1]
```diff
@@ -150,46 +150,38 @@
 
 # %% tags=["solution"]
 import numpy as np
```
Comment (glemaitre) [inline python_scripts/trees_sol_01.py:152]: I think that we can simplify the code by using some `groupby` that will avoid using the `loc` here:

```python
from matplotlib import cm

_, axs = plt.subplots(ncols=3, nrows=1, sharey=True, figsize=(12, 5))
plt.suptitle("Predicted probabilities for decision tree model", y=1.05)
plt.subplots_adjust(bottom=0.45)

for idx, (class_of_interest, data_class) in enumerate(
    data_test.groupby(by=target_test)
):
    axs[idx].set_title(f"Class {class_of_interest}")
    disp = DecisionBoundaryDisplay.from_estimator(
        tree,
        data_test,
        response_method="predict_proba",
        class_of_interest=class_of_interest,
        ax=axs[idx],
        vmin=0,
        vmax=1,
    )
    data_class.plot.scatter(
        x="Culmen Length (mm)",
        y="Culmen Depth (mm)",
        ax=axs[idx],
        marker="o",
        s=100,
        c="w",
        edgecolor="k",
    )

ax = plt.axes([0.15, 0.14, 0.7, 0.05])
plt.colorbar(
    cm.ScalarMappable(cmap="viridis"), cax=ax, orientation="horizontal"
)
_ = plt.title("Probability")
```

Comment (glemaitre) [reply python_scripts/trees_sol_01.py:152]: I feel that we have less matplotlib boilerplate code in this case since we rely on pandas directly.

**Comment by glemaitre** [comment]
I see an occurrence where we could reuse the axis from the first display, it seems more natural.

```python
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns

from sklearn.inspection import DecisionBoundaryDisplay

tab10_norm = mpl.colors.Normalize(vmin=-0.5, vmax=8.5)

palette = ["tab:blue", "tab:green", "tab:orange"]
disp = DecisionBoundaryDisplay.from_estimator(
    tree,
    data_train,
    response_method="predict",
    cmap="tab10",
    norm=tab10_norm,
    alpha=0.5,
)
sns.scatterplot(
    data=penguins,
    x=culmen_columns[0],
    y=culmen_columns[1],
    hue=target_column,
    palette=palette,
    ax=disp.ax_,
)
disp.ax_.legend(bbox_to_anchor=(1.05, 1), loc="upper left")
_ = disp.ax_.set_title("Decision boundary using a decision tree")
```

File: python_scripts/trees_sol_01.py [hunk 2]
```diff
@@ -150,46 +150,38 @@
...
+    axs[idx].set_xlabel("Culmen Length (mm)")
+    axs[idx].set_ylabel("Culmen Depth (mm)" if idx == 0 else None)
 
-ax = plt.axes([0.15, 0.04, 0.7, 0.05])
-plt.colorbar(imshow_handle, cax=ax, orientation="horizontal")
+ax = plt.axes([0.15, 0.14, 0.7, 0.05])
+plt.colorbar(
+    cm.ScalarMappable(cmap="viridis"), cax=ax, orientation="horizontal"
+)
 _ = plt.title("Probability")
```
Comment (glemaitre) [inline python_scripts/trees_sol_01.py:185]: I retouch slightly the code to iterate on the axis:

```python
from matplotlib import cm

_, axs = plt.subplots(ncols=3, nrows=1, sharey=True, figsize=(12, 5))
plt.suptitle("Predicted probabilities for decision tree model", y=1.05)
plt.subplots_adjust(bottom=0.45)

for idx, (class_of_interest, ax) in enumerate(zip(tree.classes_, axs)):
    ax.set_title(f"Class {class_of_interest}")
    DecisionBoundaryDisplay.from_estimator(
        tree,
        data_test,
        response_method="predict_proba",
        class_of_interest=class_of_interest,
        ax=ax,
        vmin=0,
        vmax=1,
    )
    ax.scatter(
        data_test["Culmen Length (mm)"].loc[target_test == class_of_interest],
        data_test["Culmen Depth (mm)"].loc[target_test == class_of_interest],
        marker="o",
        c="w",
        edgecolor="k",
    )
    ax.set_xlabel("Culmen Length (mm)")
    if idx == 0:
        ax.set_ylabel("Culmen Depth (mm)")

ax = plt.axes([0.15, 0.14, 0.7, 0.05])
plt.colorbar(
    cm.ScalarMappable(cmap="viridis"), cax=ax, orientation="horizontal"
)
_ = ax.set_title("Class probability")
```

**Review by glemaitre — APPROVED** [review]
LGTM otherwise.

**Comment by ogrisel** [comment]
If this is already there in 1.4, we can directly target `main`, no?

**Comment by ArturoAmorQ** [comment]
> If this is already there in 1.4, we can directly target main, no?

According to https://github.com/INRIA/scikit-learn-mooc/issues/789, we are currently running on v1.3. I confirmed this by running `sklearn.__version__` on the mooc.

**Review by ogrisel — APPROVED** [review]
I did another pass of rephrasing.

+1 for merging to `update-to-scikit-learn-1.6` and pushing a resync of the notebooks / exercises in that branch to check the CI.

**Review by ogrisel — APPROVED** [review]

File: python_scripts/trees_sol_01.py [hunk 3]
```diff
@@ -140,71 +140,58 @@
...
-# For example, in the plot below, the first plot on the left shows in yellow the
-# certainty on classifying a data point as belonging to the "Adelie" class. In
-# the same plot, the spectre from green to purple represents the certainty of
-# **not** belonging to the "Adelie" class. The same logic applies to the other
-# plots in the figure.
+# For example, in the plot below, the first plot on the left shows the
+# certainty of classifying a data point as belonging to the "Adelie" class. The
+# darker the color, the more certain the model is that a given point in the
+# feature space belongs to a given class the predictions. The same logic
+# applies to the other plots in the figure.
```
Comment (ogrisel) [inline python_scripts/trees_sol_01.py:151]: There is an extra "the predictions" here. Let me fix it and push a sync.
